File: src/main/java/jp/co/soramitsu/iroha2/scale/writer/BoolWriter.java
package jp.co.soramitsu.iroha2.scale.writer;

import io.emeraldpay.polkaj.scale.ScaleCodecWriter;
import io.emeraldpay.polkaj.scale.ScaleWriter;
import java.io.IOException;
import jp.co.soramitsu.iroha2.model.Bool;

public class BoolWriter implements ScaleWriter<Bool> {

  @Override
  public void write(ScaleCodecWriter writer, Bool value) throws IOException {
    writer.writeByte(value.getValue() ? 1 : 0);
  }
}

File: modules/model/src/main/kotlin/jp/co/soramitsu/iroha2/codec/writer/BoolWriter.kt
package jp.co.soramitsu.iroha2.codec.writer

import jp.co.soramitsu.iroha2.codec.ScaleCodecWriter
import jp.co.soramitsu.iroha2.codec.ScaleWriter

class BoolWriter : ScaleWriter<Boolean> {
    override fun write(wrt: ScaleCodecWriter, value: Boolean) {
        when (value) {
            false -> wrt.directWrite(0)
            true -> wrt.directWrite(1)
        }
    }
}

class BoolNullableWriter : ScaleWriter<Boolean?> {
    override fun write(writer: ScaleCodecWriter, instance: Boolean?) {
        when (instance) {
            null -> writer.directWrite(0)
            false -> writer.directWrite(1)
            true -> writer.directWrite(2)
        }
    }
}
